### PR TITLE
#3432 FridgeHeader on FridgeDetailPage A2

### LINKED ISSUE #3432: Add a header to the FridgeDetail paper area
## Is your feature request related to a problem? Please describe.

The current `FridgeDetail` page needs a header on the `Paper` area.

## Describe the solution you'd like

Currently:

image

Should be something like:

image

## Implementation

Unsure. Would be nice to have a component for the 3 different places a similar component is used. The above two and also in the `SensorEditPage`:

image

I think theres a bit of difference and only the `DetailPage` and `EditPage` will be able to share the component, where there would be a conditional to show or hide the `Cog`.

I think it would possibly be worthwhile using this same component in the `VaccinePage` also, and just also use the battery level and last sync time in the header (Since it is one to one sensor to location it's OK)

## Describe alternatives you've considered

N/A

## Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3432 

## Change summary

- Added some state for the last download status and the last download time. With these being able to be persisted as well as there soon being a `bluetooth/download` slice, I figured this could be nicely encapsulated there. Getting the last download time from the logs means you know when the last successful download was, but not the last attempt (Maybe you should only be displaying the last 'attempt' anyway), so have a status and time means we can display in the UI the last download attempt time, as well as show whether it was OK or not.
- Added a `FridgeHeader` component. This is mainly ripped out of `SensorEditPage`. Added its use in `SensorEditPage` and `FridgeDetailPage`.
- Added `LastSensorDownload` (looking for a new name if you can think of one!) component: can be given a mac address and placed anywhere to give up to date info on when the last download was etc. There ended up being a decent amount of code to handle updating itself and hooking into redux which shouldn't need to be repeated.
- Added `BlinkSensorButton` (looking for a new name if you can think of one!) component: Again can be given a mac and placed anywhere and it will handle sending the command+state updates+disabling itself when another sensor is blinking and spinning when itself is blinking.
- Used `BlinkSensorButton` on `VaccinePage`

## Testing

*Fridge Detail Page*
- [ ] Has a header component on top of the chart and the details: sensor name, battery level and last sync time are correct
- [ ] Last sync time flashes when that sensor is downloading
- [ ] Last sync icon changes when the last sync fails to be a fail icon
- [ ] Last sync time updates continuously (i.e. will say 'few seconds ago' then 'a minute ago', 'two minutes ago' etc without touching the ui at all
- [ ] when blinking a sensor, all other sensor blink buttons are disabled (can see on vaccine page)

*Sensor edit page*
- [ ] Has a header component on top of the chart and the details: sensor name, battery level and last sync time are correct
- [ ] Last sync time flashes when that sensor is downloading
- [ ] Last sync icon changes when the last sync fails to be a fail icon
- [ ] Last sync time updates continuously (i.e. will say 'few seconds ago' then 'a minute ago', 'two minutes ago' etc without touching the ui at all
- [ ] when blinking a sensor, all other sensor blink buttons are disabled (can see on vaccine page)

### Related areas to think about

- I think this `FridgeHeader` should be used on the `VaccinePage` also and remove the last sync/battery level from the flat list item. Can add the `Circle` and just use that everywhere.

### CODE REVIEW

File: src/hooks/useIntervalReRender.js
Comment (joshxg) [inline src/hooks/useIntervalReRender.js:18]: This hook is will re-render the "last download 1 min ago" component every so often which will cause a recalc the delta of the last sync time and the time now - without this, theres no re-render when time changes and if you sit on a page for 10 mins it'll just say "a few seconds ago"

File: src/pages/SensorEditPage.js
Comment (joshxg) [inline src/pages/SensorEditPage.js:5]: Main change in this component is extracting `FridgeHeader` into its own file and using it in here.

File: src/pages/VaccinePage.js
Comment (joshxg) [inline src/pages/VaccinePage.js:21]: Just added the use of `BlinkSensorButton` (Name is trademarked on this one, sorry) in this component - don't have to handle disabling/spinner etc again

File: src/reducers/VaccineReducer.js
Comment (joshxg) [inline src/reducers/VaccineReducer.js:6]: OK - I am keen for ideas of what this really should be.. 🤷 

At the moment just want to show the `WiFiIcon` or `NoWifiIcon` so two states is fine

- could make it `LAST_DOWNLOAD_OK` with a boolean.... but reasonably certain we might want to put a reason at some point - "Couldn't find sensor", "Couldn't connect with sensor". 
- So maybe it should be LAST_DOWNLOAD_ERROR and be `null` if it's OK? 
- Then I thought having "OK" was nicer than `null` to represent `OK` - ok?

**Review by joshxg — COMMENTED** [review]
To test the state changes of the `Wifi` icon/text, in `FridgeHeader`, I added into `dispatchToProps`: 

```
const syncTemperatures = () => dispatch(VaccineActions.startDownloadAllLogs());
```

And passed it to the `onPress` of the download icon button.

I also added into `downloadLogsFromSensor` in `VaccineActions`:

```
if (macAddress === 'AB:CD:EF:GH:IJ:KL') throw new Error()
```

To test the error state.

**Review by katherine-sussol — APPROVED** [review]
Looks good!

Names sound pretty reasonable to me 🤷

**Comment by joshxg** [comment]
Thanks @katherine-sussol ❤️ 

I actually branched this from https://github.com/openmsupply/mobile/pull/3445 so will wait for that to get a review before merging this.

I also merged into develop (and then back down into this) the changes you and Mark made to the `VaccineActions/Reducer` etc so hopefully it's still all good 🤣

File: src/reducers/VaccineReducer.js
Comment (mark-prins) [reply src/reducers/VaccineReducer.js:6]: The current enum meets requirements, so I think all good to have as is.
We can extend when new scenarios come up - to your point, Kat, we could have `OK` and `everything else` - so we show the ok icon if the status is ok and an error state otherwise. 
What sort of error conditions could there be? ( I see little value in second guessing possible states.. all I've seen is 'sensor disconnected' / 'connection error' sort of things. )

Also, why is it a wifi icon? these are bluetooth sensors! ( this has already confused me )

File: src/selectors/vaccine.js
Comment (mark-prins) [inline src/selectors/vaccine.js:38]: or `status !== 'OK'` to allow for other error states?

**Review by mark-prins — APPROVED** [review]
Not that it needs a second approval. Just clicking Approve because it looks marvellous

**Comment by mark-prins** [comment]
I am getting `Failed prop type: Invalid prop 'lastDownloadTime' of value 'Thu Feb 04 2021 10:26:56 GMT+1200 (+12)' supplied to 'LastSensorDownloadComponent', expected one of [null,null].` with this branch. I think merge and when all the branches are merged I'll recheck.

**Comment by joshxg** [comment]
Proptype fix: 56901c5

File: src/selectors/vaccine.js
Comment (joshxg) [reply src/selectors/vaccine.js]: Nice! 
1f417e7

**Pull request merged**
